### PR TITLE
Workspace migration

### PR DESCRIPTION
**Pull request by jolampi**
- [x] #519 
- [x] #531
- [x] #572 
- [x] #574

### CODE REVIEW

File: src/actions/moveExtensionDataPath.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    const { resources, tmc, workspaceManager } = actionContext;
+
+    const activeCourse = workspaceManager.activeCourse;
+    if (activeCourse) {
+        const exercisesToClose = workspaceManager
+            .getExercisesByCourseSlug(activeCourse)
+            .filter((x) => x.status === ExerciseStatus.Open)
+            .map((x) => x.exerciseSlug);
+
+        // Close exercises without writing the result to "reopen" them with refreshLocalExercises
```
Comment (sebazai) [inline src/actions/moveExtensionDataPath.ts:31]: Should wait for eventListener before opening, maybe.

File: src/actions/refreshLocalExercises.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        Logger.warn("Failed to determine closed status for exercises, defaulting to open.", e);
+        return [];
+    });
+    const closedExercises = new Map(
+        (closedExercisesResult.val ?? []).map((x) => [x["course-slug"], new Set(x.exercises)]),
+    );
+    const workspaceExercises = langsExercises.map<WorkspaceExercise>((x) => {
+        const exercisePath = x["exercise-path"];
+        const exerciseSlug = x["exercise-slug"];
+        const courseSlug = nth(exercisePath.split(path.sep), -2) ?? "null";
```
Comment (sebazai) [inline src/actions/refreshLocalExercises.ts:43]: tmc-langs-rust could handle course-slug as well

File: src/actions/user.ts [hunk 3]
```diff
@@ -151,9 +158,13 @@ export async function testExercise(
         template: { templateName: "test-result", ...data, pasteLink: "" },
         messageHandler: async (msg: { type?: string; data?: { [key: string]: unknown } }) => {
             if (msg.type === "submitToServer" && msg.data) {
```
Comment (sebazai) [inline src/actions/user.ts:160]: remove msg.data

File: src/actions/user.ts [hunk 4]
```diff
@@ -292,9 +306,7 @@ export async function submitExercise(
     });
     temporaryWebviewProvider.addToRecycables(temp);
 
-    const courseData = userData.getCourseByName(exerciseDetails.val.course) as Readonly<
-        LocalCourseData
-    >;
+    const courseData = userData.getCourseByName(exercise.exerciseSlug) as Readonly<LocalCourseData>;
```
Comment (sebazai) [inline src/actions/user.ts:309]: ?

File: src/actions/user.ts [hunk 5]
```diff
@@ -575,15 +567,6 @@ export async function removeCourse(actionContext: ActionContext, id: number): Pr
         Logger.error(message, closeResult.val);
         showError(message);
     }
-    const exercises = workspaceManager.getAllExerciseDataByCourseName(course.name);
-    const missingIds = exercises
-        .filter((e) => e.status === ExerciseStatus.MISSING)
-        .map((e) => e.id);
-    Logger.log(`Removing ${missingIds.length} exercise data with Missing status`);
-    workspaceManager.deleteExercise(...missingIds);
-    delSync(path.join(resources.workspaceFolderPath, course.name, ".code-workspace"), {
```
Comment (sebazai) [inline src/actions/user.ts:584]: Reconsider deleting .code-workspace file if course is removed by user.

File: src/actions/webview.ts [hunk 6]
```diff
@@ -75,28 +74,16 @@ export async function displayLocalCourseDetails(
...
-    };
-
     const mapStatus = (
         exerciseId: number,
-        status: ConfigTypes.ExerciseStatus,
+        status: ExerciseStatus,
         expired: boolean,
     ): UITypes.ExerciseStatus => {
         switch (status) {
-            case ConfigTypes.ExerciseStatus.CLOSED:
```
Comment (sebazai) [inline src/actions/webview.ts:96]: Does tmc-langs handle missing on disk exercises?

File: src/actions/moveExtensionDataPath.ts (hunk 1 above)
Comment (jolampi) [reply src/actions/moveExtensionDataPath.ts:31]: The move may actually work without running that piece of code. Worth a try.

File: src/api/tmc.ts [hunk 7]
```diff
@@ -43,6 +41,15 @@ import {
     TestResults,
 } from "./types";
 
+const API_CACHE_LIFETIME = 5 * 60 * 1000;
+const CLI_PROCESS_TIMEOUT = 2 * 60 * 1000;
```
Comment (sebazai) [inline src/api/tmc.ts:45]: Correct place would be constants.ts, why were these moved?

File: src/api/tmc.ts [hunk 8]
```diff
@@ -258,11 +304,97 @@ export default class TMC {
...
+    ): Promise<Result<Array<{ id: number }>, Error>> {
+        return this._executeLangsCommand(
+            { args: ["check-exercise-updates"], core: true },
+            createIs<Array<{ id: number }>>(),
+            { forceRefresh: options?.forceRefresh, key: TMC._exerciseUpdatesCacheKey },
+        ).then((res) => (res.err ? res : Ok(res.val.data)));
+    }
+
+    /**
+     * @deprecated - Migrate to `downloadExercises`
```
Comment (sebazai) [inline src/api/tmc.ts:397]: Should this be removed then?

File: src/migrate/index.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    tmc: TMC,
+): Promise<Result<void, Error>> {
+    const memento = context.globalState;
+
+    try {
+        const migratedExtensionSettings = await migrateExtensionSettings(memento);
+        const migratedSessionState = migrateSessionState(memento);
+        const migratedUserData = migrateUserData(memento);
+
+        // Workspace data migration - this one is a bit more tricky so do it last.
```
Comment (sebazai) [inline src/migrate/index.ts:31]: Is this done?

**Comment by sebazai** [comment]
Migration seems to move all exercise folders from earlier specified location in version 1.3.4, e.g. in my case G:\tmcdata\ -> %appdata_local%\tmc\vscode\.

This location should not change if data path has been set by user in earlier version (i.e. exercises are not in default vscode globalStorage)

**Comment by sebazai** [comment]
In my case I had around 1000 different exercises and the migration from version 1.3.4 data to 2.0.0 took quite a while. A progress bar of the migration progress is necessary in these situations.

**Comment by sebazai** [comment]
Some failure in activating the extension during migration.
`Activating extension 'moocfi.test-my-code' failed: ENOENT: no such file or directory, open 'c:\Users\xxxx\AppData\Roaming\Code\User\globalStorage\moocfi.test-my-code\workspaces\mooc-ohjelmointi-21.code-workspace'.`

Steps to reproduce:
- Install tmc-vscode 1.3.4, add few courses and download exercises, set some to open/close status.
- Uninstall tmc-vscode extension 1.3.4
- Start this branch in production mode

File: src/migrate/index.ts (hunk 9 above)
Comment (jolampi) [reply src/migrate/index.ts:31]: It is done.

File: src/api/tmc.ts (hunk 8 above)
Comment (jolampi) [reply src/api/tmc.ts:397]: This function is still used in some tests, e.g. downloading an old submission requires an exercise to be present in the first place. This requires larger changes as the new `downloadExercises` uses external (langs) configuration to determine where downloads should go.

File: src/api/workspaceManager.ts [hunk 10]
```diff
@@ -369,202 +171,167 @@ export default class WorkspaceManager {
...
+            }
+        });
+
+        e.removed.forEach((removed) => {
+            const exercise = this._exercises.find((x) => x.uri === removed.uri);
+            if (exercise) {
+                exercise.status = ExerciseStatus.Closed;
+            }
+        });
+
```
Comment (sebazai) [inline src/api/workspaceManager.ts:279]: Collapse would be great to have as earlier for less cluttered code-workspace file:
https://github.com/rage/tmc-vscode/blob/d572e0218fb974bcb73da1e8ddb8106a0e5424be/src/init/index.ts#L83

File: src/api/workspaceManager.ts [hunk 11]
```diff
@@ -369,202 +171,167 @@ export default class WorkspaceManager {
...
+        );
+        if (!success) {
+            Logger.error("Replace operation failed.");
+            Logger.debug("Failed with folders:", ...correctStructure);
+        }
+
+        return success ? Ok.EMPTY : Err(new Error("Failed to refresh active workspace."));
+    }
+
+    private _onDidChangeWorkspaceFolders(e: vscode.WorkspaceFoldersChangeEvent): void {
```
Comment (sebazai) [inline src/api/workspaceManager.ts:260]: For some reason this doesn't fire if I manually remove folders from the workspace. That is: Right click a root folder and "Remove folder from Workspace"
